### PR TITLE
[Feature] Add native Qwen3-Reranker support and SM87 compute capability

### PR DESCRIPTION
**Pull request by thomas-hiddenpeak**
# What does this PR do?

This PR adds two features:

## 1. Native Qwen3-Reranker Support

Adds support for **Qwen3-Reranker** models (generative reranker). Unlike traditional classifiers that use a classification head, Qwen3-Reranker computes relevance scores using yes/no token logits from a causal language model.

### How it works:
1. Input is formatted using ChatML template with a system prompt asking for yes/no judgment
2. The model generates logits for all tokens
3. We extract logits at the **last position** for yes/no tokens (YES=9693, NO=2152)
4. Score = `P(yes) = softmax([logit_no, logit_yes])[1]`

### Changes:
- Added `ModelType::ListwiseReranker` variant in `backends/core/src/lib.rs`
- Implemented `lm_head` with tied embeddings support in `qwen3.rs` and `flash_qwen3.rs`
- Added ChatML template system (`core/src/templates.rs`)
- Model detection: Qwen3 without `1_Pooling/config.json` → ListwiseReranker
- Skip sigmoid for ListwiseReranker (already returns probability)

## 2. SM87 Compute Capability Support (Jetson Orin)

Added support for NVIDIA Jetson Orin (compute capability 8.7) in `backends/candle/src/compute_cap.rs`.

## Testing

Tested on both CPU and GPU (Jetson AGX Orin SM87):

```bash
# Build with SM87 support
CUDA_COMPUTE_CAP=87 cargo build --release -p text-embeddings-router -F http -F candle-cuda

# Reranker test
curl http://localhost:8090/rerank \
  -H "Content-Type: application/json" \
  -d '{"query": "What is Deep Learning?", "texts": ["Deep Learning is a branch of machine learning", "The weather is nice today"]}'

# Response: [{"index":0,"score":0.47852883},{"index":1,"score":0.0000104515475}]
```

## Related

- Related to #730 (seq-cls approach) - this PR provides an alternative approach using the original generative model
- [Qwen3-Reranker Model Card](https://huggingface.co/Qwen/Qwen3-Reranker-0.6B)

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/text-embeddings-inference/blob/main/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a GitHub issue or the forum? Related to #730
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

@Narsil OR @alvarobartt OR @kozistr

### CODE REVIEW

**Comment by michaelfeil** [comment]
@thomas-hiddenpeak can you refactor into a tiny pr that:
- supports sm87


- listwise reranker: I think the design of TEI is quite elegant: backend runs though the model, classifier optionally runs heads. No postprocessing is happening. TEI could have easily introduced "reranker", but its just a classification model. 

Listwise reranker functionality should not be introdouced as backend, but as "core/router" level feature, if any and implemented on http/grpc level.

**Comment by thomas-hiddenpeak** [comment]
> @thomas-hiddenpeak can you refactor into a tiny pr that:
> 
> * supports sm87
> * listwise reranker: I think the design of TEI is quite elegant: backend runs though the model, classifier optionally runs heads. No postprocessing is happening. TEI could have easily introduced "reranker", but its just a classification model.
> 
> Listwise reranker functionality should not be introdouced as backend, but as "core/router" level feature, if any and implemented on http/grpc level.

Hi,

Thank you very much for the feedback and for explaining TEI's design philosophy so clearly! 🙏

I completely understand and agree: the backend should remain focused solely on model execution + optional classification heads, with no post-processing inside.

Could you please provide more details on how you would prefer to handle this kind of "special output" logic (e.g., extracting specific token logits and computing a score) at the core/router level?

- What kind of abstraction or pattern do you have in mind?
- Are there any existing parts of the codebase (in router, handlers, etc.) that deal with similar output transformations I could reference?

I'm ready to refactor the PR to move everything accordingly — just need your guidance on the right place and way to do it to keep consistent with the project's elegance.

Thanks a lot for your time!

**Comment by thomas-hiddenpeak** [comment]
> @thomas-hiddenpeak can you refactor into a tiny pr that:
> 
> * supports sm87
> * listwise reranker: I think the design of TEI is quite elegant: backend runs though the model, classifier optionally runs heads. No postprocessing is happening. TEI could have easily introduced "reranker", but its just a classification model.
> 
> Listwise reranker functionality should not be introdouced as backend, but as "core/router" level feature, if any and implemented on http/grpc level.

Just to clarify — I actually submitted a standalone PR for SM87 support earlier:
#467 

It has been tested and used in production on Jetson Orin devices for quite some time without issues. However, it hasn’t been merged yet, and I realize I might not fully understand the community’s contribution guidelines or review expectations.

If you have a moment, could you kindly let me know:

What aspects of that PR need improvement?
Are there specific style, testing, or architectural concerns I should address?
I’d really appreciate your guidance so I can update it accordingly and help get SM87 (and potentially SM89) support merged cleanly.

Thanks again for your time and support!

File: backends/candle/src/models/flash_qwen3.rs [hunk 1]
```diff
@@ -502,6 +531,80 @@ impl FlashQwen3Model {
...
+            let no_logit = last_logits.i(NO_TOKEN_ID)?.to_dtype(DType::F32)?.to_scalar::<f32>()?;
+
+            // Compute P(yes) using softmax over [no, yes]
+            let max_logit = yes_logit.max(no_logit);
+            let yes_exp = (yes_logit - max_logit).exp();
+            let no_exp = (no_logit - max_logit).exp();
+            let p_yes = yes_exp / (yes_exp + no_exp);
+
+            scores.push(vec![p_yes]);
+        }
```
Comment (kozistr) [inline backends/candle/src/models/flash_qwen3.rs:602]: I guess that maybe we could vectorize this by creating the last indices tensor first and do the rest!

File: backends/candle/src/models/flash_qwen3.rs [hunk 2]
```diff
@@ -502,6 +531,80 @@ impl FlashQwen3Model {
 
         Ok((pooled_embeddings, raw_embeddings))
     }
+
+    /// Forward pass for ListwiseReranker
+    /// Returns P(yes) scores for reranking
+    pub fn forward_for_rerank(&self, batch: Batch) -> Result<Tensor> {
```
Comment (kozistr) [inline backends/candle/src/models/flash_qwen3.rs:537]: As mentioned [here](https://github.com/huggingface/text-embeddings-inference/pull/795#issuecomment-3744577664), 

Maybe we could just create a classification head like other implementations instead of introducing a forward function for the re-ranker!

Here's a brief pseudo-code!
```
...

impl Qwen3ClassificationHead {
   ... load() {
       /// load `lm_head` here 
   }

   fn forward(self, hidden_states: &Tensor) -> Result<Tensor> {
       /// move forward_for_rerank logic here
   }
}

...

    /// predict part
    fn predict(&self, batch: Batch) -> Result<Tensor> {
        match &self.classifier {
            None => candle::bail!("`predict` is not implemented for this model"),
            Some(classifier) => {
                let (pooled_embeddings, _) = self.forward(batch)?;
                let pooled_embeddings =
                    pooled_embeddings.expect("pooled_embeddings is empty. This is a bug.");
                classifier.forward(&pooled_embeddings)
            }
        }
    }
```

If we need to keep two types of classifiers (e.g. listwise reranker, binary classifier), we could make an enum for the classifiers and branch by the weight name or something.

I might be wrong, so please feel free to leave any feedback or comments!

**Comment by kozistr** [comment]
Hi! I've just opened [PR](https://github.com/huggingface/text-embeddings-inference/pull/835) to support Qwen3-Reranker!

Please feel free to share your thoughts :)

**Comment by thomas-hiddenpeak** [comment]
Thanks for all the feedback! I've split this PR into focused parts:

- **SM87 + SM110 support** → #844 (already submitted)
- **Qwen3-Reranker** → @kozistr has opened a cleaner implementation in #835, so I'll defer to that PR

Closing this one as superseded. Thanks everyone for the guidance!
